Ibn 613 add new temporary storage

Maw-Jung opened the pull request.

mkhl reviewed (changes requested):
Ah, so that’s what you were working on.

First of all, please expand your commit messages.
When finding changes like these in the history, I’d like to know

* which problem it is trying to solve
* how it’s solving that problem
* why it’s solving the problem in this way,
  including why other ways were disregarded

I know you experimented with service rankings,
and I’ll assume you tried replacing PersistentInbox with TemporaryInbox through that?
If so, the reason it didn’t help is probably
that PersistentInbox registers itself as a DiscoveryListener,
bypassing OSGi.
Is that what you tried?

Considering alternatives,
I think I’d try to add and use an alternative Storage (and StorageService) implementation
that just keep things in memory, like this `temporaryStorage` does.
Maybe call is `TemporaryStorage`, `VolatileStorage`, `HeapStorage`, `InMemoryStorage` or something.
Then add a `@Property` to that StorageService
and use the configuration to set a reference target filter on PersistentInbox
so it uses the volatile storage instead of MapDB.

Have you tried something like that?
[Here’s a blog post that can help get you started](https://thecarlhall.wordpress.com/2010/06/06/set-osgi-service-reference-target-through-configuration/) if you haven’t.
Something like `persistentInbox:storageService.target` should work
as the configuration property I think.

mkhl commented:
Oh, there even _is_ a volatile storage already, just in a test bundle: `org.eclipse.smarthome.test.VolatileStorage` (and it correctly uses a `ConcurrentHashMap`)

Maw-Jung commented:
> I know you experimented with service rankings,
> and I’ll assume you tried replacing PersistentInbox with TemporaryInbox through that?
> If so, the reason it didn’t help is probably
> that PersistentInbox registers itself as a DiscoveryListener,
> bypassing OSGi.
> Is that what you tried?

Indeed, that was the last conclusion i wrote to you about and the reason why i decided to replace the PersistentInbox.

mkhl commented:
Great, then we know why exactly that approach doesn’t work, and we have a possible alternative to just replacing the upstream module wholesale.

Commit:
IBN-613: Add temporary storage to preserve sd card life

The PersistentInbox used the mapbd storage to store all discovery results,
which put wear on the sd card with the writes to it. To prolong the sd card
lifetime, the storage for inbox entries should be in ram. The PersistentInbox
is left, even tho the name is kinda missleading, because we want to obtain as
much compability with esh. A second inbox, which is used through higher order
ranking doesn't work either, because the Inbox itself registers as a Discovery-
Listener. Thus the inbox would be doubled. Through an additional storage
implementation, copied from the test bundle into a new bundle, we can now use
this TemporaryStorage in place of the mapdb storage.

To actually use this new storage implementation in PersistentInbox,
add the following to your configuration:

    org.eclipse.smarthome.config.discovery.internal.PersistentInbox:StorageService.target=(type=temporaryStorage)

mkhl commented on bundles/storage/org.eclipse.smarthome.storage.temporary/OSGI-INF/org.eclipse.smarthome.storage.temporary.TemporaryStorageService.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Doesn’t this file get automatically generated?

Maw-Jung replied:
It does and is tracked in other bundles, what do you recommend we do?

mkhl replied:
It is not tracked but ignored in ESH and in newer bundles. `git rm` the xml files, add a `.gitignore` file containing `*.xml` to OSGI-INF.

mkhl replied:
Oh, that reminds me: When starting liteserver, `ManualDiscoveryResource` throws errors about already being registered. That’s exactly the kind of error this would produce. Have you checked in an OSGi service XML file despite them explicitly being ignored in that directory? If so, you should remove it.

aknrdureegaesr replied:
For what it's worth: I saw that duplicate error today, too, in a rather virgin dev environment (based on `testing`).

```

org.osgi.service.component.ComponentException: The component name 'org.eclipse.smarthome.binding.hue.internal.discovery.ManualDiscoveryResource' has already been registered by Bundle 84 (org.eclipse.smarthome.binding.hue) as Component of Class org.eclipse.smarthome.binding.hue.internal.discovery.ManualDiscoveryResource
	at ...
```

Maw-Jung replied:
Jeah, i did in https://github.com/AlJungOrg/smarthome/pull/29/files, will remove it

mkhl commented on bundles/storage/org.eclipse.smarthome.storage.temporary/OSGI-INF/temporarystorage.xml:
```diff
@@ -0,0 +1,23 @@
+	SPDX-License-Identifier: EPL-2.0
+
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" activate="activate" deactivate="deactivate" immediate="true" name="org.eclipse.smarthome.storage.temporary">
```
Isn’t this mostly a duplicate of the other service descriptor? Do we need both? Don’t they get auto-generated?

Maw-Jung replied:
Right, this one was modified by me and the other one got generated, so i removed this one.

mkhl commented on bundles/storage/org.eclipse.smarthome.storage.temporary/src/main/java/org/eclipse/smarthome/storage/temporary/TemporaryStorageService.java:
```diff
@@ -0,0 +1,53 @@
+ * @author Thomas.Eichstaedt-Engelen - Initial Contribution and API
+ * @author MAW - moved to new bundle and renamed
+ */
+@Component(
```
A single line is fine for this annotation :-)

Maw-Jung replied:
done

mkhl commented on bundles/storage/org.eclipse.smarthome.storage.temporary/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,25 @@
+Bundle-Vendor: Eclipse.org/SmartHome
+Bundle-Version: 0.10.0.qualifier
+Import-Package: 
+ com.google.gson,
```
I don’t see GSON being used anywhere in this bundle. Can we get rid of the imports we don’t require?

Maw-Jung replied:
done

mkhl commented on bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/PersistentInbox.java:
```diff
@@ -646,7 +646,7 @@ protected void unsetManagedThingProvider(ManagedThingProvider thingProvider) {
     }
 
-    @Reference(policy = ReferencePolicy.DYNAMIC)
+    @Reference(policy = ReferencePolicy.DYNAMIC, target="(type=temporaryStorage)")
```
Nice. Now see if we can’t do the same thing without modifying the ESH sources, just using configuration.

A Component can be configured via its `configurationPid`. If unset, it is set to the same value as the component `name`. If _that_ is unset, it uses the fully qualified name of the annotated class.

What I’ve read leads me to believe that we can use that to set a target for this reference using a configuration parameter like this:
```
org.eclipse.smarthome.config.discovery.internal.PersistentInbox:storageService.target
```

Does that work?

pwolfart commented:
title?

mkhl commented:
@Maw-Jung 

I have

* reverted the change to the `@Reference` in `PersistentInbox`
* removed all commits that were later reverted
* squashed the remaining commit and its fixup commit
* added the new storage to the storage pom so maven actually builds it
* added a note about the required configuration change to the commit message

You will need to update AlJungOrg/liteserver#959 such that it

* updates the eclipse smarthome dependency to a version including this change
* turns the “fixup” commit into an actual fixup instead of a commit
* adds a new configuration file in `distribution/openhabhome/conf/services/` with an appropriately name ending in `.cfg` and containing this line:

        org.eclipse.smarthome.config.discovery.internal.PersistentInbox:StorageService.target=(type=temporaryStorage)
